### PR TITLE
Stowrs error reporting and export adapter changes (fixed cla)

### PR DESCRIPTION
**Pull request by red1408**
- Deprecate peer_dicomweb_addr peer_dicomweb_stow_path
- Jetty client http2 implementation of stow-rs used by c-store service and export adapter. Main difference - it can receive (error) response while transmitting data stream.
- updated to java 11. Initially because I tried to use native new java http client. Later left as is, since there seem to be no drawbacks.

Baked both pull requests into one because export adapter got broken pipe errors while trying to use old stowrs implementation (exactly same as #5), but when I tried to replace client with Jetty to actually see the error, it just worked. Doubt it's worth looking in too deeply, since that's what I was going to do next anyway.

Note: Jetty client implements only stow-rs (POST with binary data). I don't see particularly good reason to over-complicate GET requests with use of low level api.

### CODE REVIEW

File: dicom_util/src/main/java/com/google/cloud/healthcare/Util.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.google.cloud.healthcare;
```
Comment (danielbeaudreau) [inline dicom_util/src/main/java/com/google/cloud/healthcare/Util.java]: I'm not sure it's worth introducing a new class for just this one method. I think duplicating it is fine given it's just one line.

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      }
+    } catch (Exception e) {
+      if (e instanceof DicomWebException) {
+        throw (DicomWebException) e;
+      }
+      throw new DicomWebException(e);
+    }
+  }
+
+  private static class DataStream {
```
Comment (danielbeaudreau) [inline dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: The Jetty implementation doesn't provide any utilities for streaming data? I would have thought there would be something we could use instead of having to write these lower level classes ourselves

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java [hunk 3]
```diff
@@ -120,8 +120,8 @@ public static void main(String[] args) throws IOException, GeneralSecurityExcept
 
     Subscriber subscriber = null;
     try {
-      SubscriptionName subscriptionName =
-          SubscriptionName.of(Flags.projectId, Flags.subscriptionId);
+      ProjectSubscriptionName subscriptionName =
```
Comment (danielbeaudreau) [inline export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: Why the change of class here? Java 11 compatibility?

**Comment by red1408** [comment]
> Any chance you could squash this PR so it's fewer commits, rather than 91?

Right, force-pushed branch with only one commit. I'm not sure why it ends up showing 91, when created directly from master (at least changes are actually all the same).

File: dicom_util/src/main/java/com/google/cloud/healthcare/Util.java (hunk 1 above)
Comment (red1408) [reply dicom_util/src/main/java/com/google/cloud/healthcare/Util.java]: I just really dislike code duplication. There could be more clients later too (well, there were more, during testing/working it out phase).

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java (hunk 2 above)
Comment (red1408) [reply dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: No, there doesn't seem to be a class for this.

There is whole high level Jetty client though (HttpClient class as opposed to low level Http2Client used by me). And it almost works, but ends up violating flow control during send more often than not (sending more data than allowed by WINDOW_UPDATE). I'm not sure how to fix this in high level client (If I did, I wouldn't need the low level implementation).

While current low level implementation just seems to work (I send next frame after callback succeeds, so no flow control errors there).

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java (hunk 3 above)
Comment (red1408) [reply export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: More specifically changes in 'com.google.cloud:google-cloud-pubsub:1.89.0' (I couldn't keep old version due to dependencies conflict).
From what I understand it's exactly same thing, but renamed.

File: dicom_util/src/main/java/com/google/cloud/healthcare/Util.java (hunk 1 above)
Comment (danielbeaudreau) [reply dicom_util/src/main/java/com/google/cloud/healthcare/Util.java]: Ok, can we rename this class to StringUtil instead?

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+          requestFields);
+      HeadersFrame headersFrame = new HeadersFrame(request, null, false);
+
+      // Prepare the listener to receive the HTTP response frames.
+      final StringBuilder resultBuilder = new StringBuilder();
+      final CompletableFuture<Integer> responseCodeFuture = new CompletableFuture<>();
+      final CompletableFuture<Boolean> doneFuture = new CompletableFuture<>();
+      Stream.Listener responseListener = new Stream.Listener.Adapter() {
+        @Override
+        public void onReset(Stream stream, ResetFrame frame) {
```
Comment (danielbeaudreau) [inline dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: When does onReset get called?

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      SslContextFactory sslContextFactory = new SslContextFactory.Client();
+      client.addBean(sslContextFactory);
+      client.start();
+
+      HttpURI uri = new HttpURI(serviceUrlPrefix + "/" + Util.trim(path));
+
+      FuturePromise<Session> sessionPromise = new FuturePromise<>();
+      client.connect(sslContextFactory, new InetSocketAddress(uri.getHost(), CONNECT_PORT),
+          new ServerSessionListener.Adapter(), sessionPromise);
+      Session session = sessionPromise.get(5, TimeUnit.SECONDS);
```
Comment (danielbeaudreau) [inline dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: 5 seconds might be too low for very large files. can we increase this limit to like 10 minutes?

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java (hunk 2 above)
Comment (danielbeaudreau) [reply dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: Ok, can you please add a comment explaining why we need the low level method? (flow control error) Thanks

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java (hunk 3 above)
Comment (danielbeaudreau) [reply export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: I see, could you please test manually that the export adapter still works after this change?
We should probably have an integration test for the export adapter if we are making changes to it, but it's not really a high priority task right now.

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java [hunk 6]
```diff
@@ -79,22 +83,18 @@ public static void main(String[] args) throws IOException, GeneralSecurityExcept
...
-      } else {
-        requestFactory = createHttpRequestFactory(null);
-      }
-      DicomWebClient exportDicomWebClient =
-          new DicomWebClient(requestFactory, flags.peerDicomwebAddr);
+      boolean isLegacyAdress = flags.peerDicomwebAddress.isEmpty();
+      String peerDicomwebAddress =
+          isLegacyAdress ? flags.peerDicomwebAddr : flags.peerDicomwebAddress;
+      String peerDicomwebStowpath = isLegacyAdress ? flags.peerDicomwebStowPath : "studies";
+      IDicomWebClient exportDicomWebClient =
```
Comment (danielbeaudreau) [inline export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: Would any post request succeed without this flag set to true? How would we get any credentials for the request?

**Comment by danielbeaudreau** [comment]
What do you think about moving the jetty stow implementation into the default dicomwebclient? Because cmove also needs stow, however it will be using the non-jetty implementation. Ideally all callers would use the same jetty implementation.

File: dicom_util/src/main/java/com/google/cloud/healthcare/Util.java (hunk 1 above)
Comment (red1408) [reply dicom_util/src/main/java/com/google/cloud/healthcare/Util.java]: Ok

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java (hunk 3 above)
Comment (red1408) [reply export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: I already tested it manually, that's the point where I had to replace old stowrs implementation with jetty one.

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java (hunk 5 above)
Comment (red1408) [reply dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: It's timeout to establish the session, it doesn't affect data transmission.

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java (hunk 6 above)
Comment (red1408) [reply export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: I assumed sending request without auth header was the intention here, but after looking deeper it doesn't seem to be right.

Though I'm also not sure under which conditions this flag was useful.
credentials are obtained as:

```
GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
if (!flags.oauthScopes.isEmpty()) {
      credentials = credentials.createScoped(Arrays.asList(flags.oauthScopes.split(",")));
}
```

The flag was used  like this:

```
// DicomWeb client for sink of DICOM.
// Use application default credentials for HTTP requests of DICOM sink, if specified by flag.
      HttpRequestFactory requestFactory = null;
      if (flags.useGcpApplicationDefaultCredentials) {
        requestFactory = createHttpRequestFactory(credentials);
      } else {
        requestFactory = createHttpRequestFactory(null);
      }
      DicomWebClient exportDicomWebClient =
          new DicomWebClient(requestFactory, flags.peerDicomwebAddr);
```

... So, assuming comments here are correct, either way it's application default credentials, with scopes being the only difference.

Can I just remove or ignore this flag?

**Comment by red1408** [comment]
> What do you think about moving the jetty stow implementation into the default dicomwebclient? Because cmove also needs stow, however it will be using the non-jetty implementation. Ideally all callers would use the same jetty implementation.

C-move instructs receiver to execute c-store against c-move target. There is no stow involved.
I could make a composite IDicomwebclient implementation when it becomes necessary, but at current point there is no service using all 3 dicomweb methods.

File: dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java (hunk 4 above)
Comment (red1408) [reply dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java]: When error response comes during post body transmission. Frame sequence for error is Header  -> Data (not end stream)  -> Reset.
For normal response it's Header  -> Data (end stream).

File: export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java (hunk 6 above)
Comment (danielbeaudreau) [reply export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java]: From the comment on Flags.java: 
"If true, uses GCP Application Credentials to when sending HTTP requests to peer. This is useful if the peer DICOMWeb endpoint is running in GCP."
So I realized that this is for non-GCP dicomWeb backends. So it's fine to leave as-is.

**Comment by danielbeaudreau** [comment]
Looks good to me. Can you please add a follow up PR to delete the old STOW implementation in the non-jetty DicomWebClient, and instead throw an unimplemented error?
